### PR TITLE
day of week cleanup

### PR DESCRIPTION
**Pull request by simonbyrne**
* Make `DayOfWeek` an `Enum`, deprecate `dayofweek`. (#19210)
* Provide optional argument to `firsdayofweek`/`lastdayofweek` (#19208)
* A few misc fixes.

### CODE REVIEW

File: test/dates/adjusters.jl [hunk 1]
```diff
@@ -220,7 +220,7 @@ dt = Dates.Date(2014,5,21)
 @test Dates.tonext(dt,Dates.Mon) == Dates.Date(2014,5,26)
 @test Dates.tonext(dt,Dates.Tue) == Dates.Date(2014,5,27)
 # No dayofweek function for out of range values
-@test_throws KeyError Dates.tonext(dt,8)
+#@test_throws KeyError Dates.tonext(dt,8)
```
Comment (quinnj) [inline test/dates/adjusters.jl]: Are you planning on adding this back in?

**Comment by quinnj** [comment]
Looking good.

File: test/dates/adjusters.jl (hunk 1 above)
Comment (simonbyrne) [inline test/dates/adjusters.jl]: No, it's going to be removed (`tonext` no longer takes integer arguments).

File: test/dates/adjusters.jl [hunk 2]
```diff
@@ -219,8 +219,6 @@ dt = Dates.Date(2014,5,21)
 @test Dates.tonext(dt,Dates.Sun) == Dates.Date(2014,5,25)
 @test Dates.tonext(dt,Dates.Mon) == Dates.Date(2014,5,26)
 @test Dates.tonext(dt,Dates.Tue) == Dates.Date(2014,5,27)
-# No dayofweek function for out of range values
-#@test_throws KeyError Dates.tonext(dt,8)
```
Comment (femtotrader) [inline test/dates/adjusters.jl]: A test with

```julia
@test Dates.firstdayofweek(Dates.DateTime(2013,12,24), firstday=Monday) == Dates.DateTime(2013,12,23)
@test Dates.firstdayofweek(Dates.DateTime(2013,12,24), firstday=Sunday) == Dates.DateTime(2013,12,22)
```

could be a good idea

and a similar test with `lastdayofweek`

Something like

````julia
@test Dates.lastdayofweek(Dates.DateTime(2013,12,24), firstday=Monday) == Dates.DateTime(2013,12,29)
@test Dates.lastdayofweek(Dates.DateTime(2013,12,24), firstday=Sunday) == Dates.DateTime(2013,12,28)
```

Comment (simonbyrne) [inline test/dates/adjusters.jl]: Good call, done.

**Comment by simonbyrne** [comment]
Okay, hopefully this should be ready. One question is whether the optional arguments to `firstdayofweek`/`lastdayofweek` should be keyword or positional? (I've gone with positional at the moment).

**Comment by nalimilan** [comment]
`dayname` and `dayabbr` use a keyword argument `locale`, which is conceptually similar. So maybe use a keyword argument here too?

**Comment by simonbyrne** [comment]
Also, should we deprecate `ismonday`, `istuesday`, etc? These seem like overkill.

**Comment by quinnj** [comment]
I'd be fine with either positional or keyword + deprecating `ismonday` `istuesday`; those were only there for the `Dates.tonext(dt, ::dayofweek)` calls

**Comment by simonbyrne** [comment]
Okay, this is rebased and ready to go. Any objections?

**Comment by simonbyrne** [comment]
@nanosoldier `runbenchmarks("dates", vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19212#issuecomment-261571746) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/617fc587144c20cc661eeffb6de8c04d093c4285/0ddaec6_vs_51da590/report.md). cc @jrevels

**Comment by KristofferC** [comment]
That looks like an objection :p

**Comment by jrevels** [comment]
I'd assume the regressions are from the deprecation right? If that's true, then the slowdown is expected and the benchmark should just be rewritten.

**Comment by simonbyrne** [comment]
Well, the `firstdayofweek` is legitimately slower, as I changed the function definition slightly.

**Comment by tkelman** [comment]
Do we still want to do this?

File: base/dates/query.jl [hunk 3]
```diff
@@ -20,49 +17,55 @@ const MONTHDAYS = [0,31,59,90,120,151,181,212,243,273,304,334]
...
+```
 """
-    dayabbr(dt::TimeType; locale="english") -> AbstractString
+DayOfWeek(n::Integer) = convert(DayOfWeek, mod1(n, 7))
 
-Return the abbreviated name corresponding to the day of the week of the `Date` or `DateTime`
-in the given `locale`.
 """
-dayabbr(dt::Integer;locale::AbstractString="english") = VALUETODAYOFWEEKABBR[locale][dt]
+    DayOfWeek(dt::TimeType)
```
Comment (quinnj) [inline base/dates/query.jl]: I think I'd like to still keep this lowercase `dayofweek`, it follows the more general pattern in Dates with types being proper-cased and accessors being lowercased.

File: doc/stdlib/dates.rst [hunk 4]
```diff
@@ -61,7 +61,7 @@ Dates Functions
 ---------------
 
 All Dates functions are defined in the ``Dates`` module; note that only the ``Date``, ``DateTime``, and ``now`` functions are exported;
-to use all other ``Dates`` functions, you'll need to prefix each function call with an explicit ``Dates.``, e.g. ``Dates.dayofweek(dt)``.
+to use all other ``Dates`` functions, you'll need to prefix each function call with an explicit ``Dates.``, e.g. ``Dates.DayOfWeek(dt)``.
```
Comment (quinnj) [inline doc/stdlib/dates.rst]: Looks like these docs need to be moved from dates.rst.

**Review by quinnj — CHANGES_REQUESTED** [review]
Yes, I'm still in favor of this. I don't think the previously noted performance regression should stop this; we're adding functionality that makes up for it.

**Comment by vtjnash** [comment]
Seems dead?

**Pull request closed without merging**